<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Capstone-Project---Planet-Jordan" data-toc-modified-id="Capstone-Project---Planet-Jordan-1"><span class="toc-item-num">1&nbsp;&nbsp;</span><img src="https://ga-dash.s3.amazonaws.com/production/assets/logo-9f88ae6c9c3871690e33280fcf557f33.png" alt="">Capstone Project - Planet Jordan</a></span><ul class="toc-item"><li><span><a href="#Background" data-toc-modified-id="Background-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Background</a></span></li><li><span><a href="#Problem-Statement" data-toc-modified-id="Problem-Statement-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Problem Statement</a></span></li><li><span><a href="#Executive-Summary" data-toc-modified-id="Executive-Summary-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Executive Summary</a></span><ul class="toc-item"><li><span><a href="#Data-cleaning" data-toc-modified-id="Data-cleaning-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>Data cleaning</a></span></li><li><span><a href="#EDA" data-toc-modified-id="EDA-1.3.2"><span class="toc-item-num">1.3.2&nbsp;&nbsp;</span>EDA</a></span></li><li><span><a href="#Modeling" data-toc-modified-id="Modeling-1.3.3"><span class="toc-item-num">1.3.3&nbsp;&nbsp;</span>Modeling</a></span></li><li><span><a href="#Business-analysis" data-toc-modified-id="Business-analysis-1.3.4"><span class="toc-item-num">1.3.4&nbsp;&nbsp;</span>Business analysis</a></span></li></ul></li><li><span><a href="#Data-dictionary" data-toc-modified-id="Data-dictionary-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Data dictionary</a></span></li><li><span><a href="#Data-Cleaning" data-toc-modified-id="Data-Cleaning-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Data Cleaning</a></span><ul class="toc-item"><li><span><a href="#Handling-null-values" data-toc-modified-id="Handling-null-values-1.5.1"><span class="toc-item-num">1.5.1&nbsp;&nbsp;</span>Handling null values</a></span></li></ul></li></ul></li></ul></div>

# ![](https://ga-dash.s3.amazonaws.com/production/assets/logo-9f88ae6c9c3871690e33280fcf557f33.png)Capstone Project - Planet Jordan

## Background
**Client brief**
- Our client is Planet Jordan, an international hotel group owned by billionaire sportsman and entrepreneur Michael Jordan
- They have five hotel chains with widespread distribution across Greater China
- Michael Jordan is very focused on the staycation market (i.e. domestic residents staying in a local hotel for a vacation, to feel like they are getting away from home)
- Planet Jordan have historically been very strong in this sector and run a loyalty program, “The Staycation Card” to encourage repeat business
- Recently Michael Jordan has noticed declines in demand and suspects it is because competitors are developing their own programs to take on Planet Jordan in key cities or across key chains
- He has asked his sales & marketing executives to prepare him with some insight on what is happening and some recommendations on what to do – they are due to present to him in three months and have engaged us in support
- Their sales & marketing team are visiting Shanghai and have asked for a preliminary meeting with us – they are primarily interested in how we will approach the project, but they are also looking for some preliminary insight on which cities and hotels are the most important to target in the project 

## Problem Statement
**Instructions**
- Produce a deck to present to the Planet Jordan Sales & Marketing team 
- Your presentation should focus on how you are going to address this question and what analyses your team will perform.  You should also provide an initial point of view on which cities and hotels should be the focus of the project
- You have been provided with some supporting excel data entitled ‘Planet Jordan Hotels_Raw Data’ – this extract will be helpful for your point of view and you may use it however we wish. You may also choose to reference other external data, but note that we are most interested in how you structure your thinking about the problem rather than your ability to regurgitate existing research 
- Add a predictive analytics by forecasting demand for rooms.
- There will be no opportunity to ask questions before your final presentation, so please call out any assumptions you have made at that time

## Executive Summary
### Data cleaning
After loading in the data, we took a quick look at the data to get a general feel of the data that we have to be dealing with. First we changed the column names to lower case for easier handling and then checked for null values, followed by checking for unusual values in both categorical and numerical columns. We found some errors and corrected them. Next we handled the null values by dropping most of them since it didn't affect the data much. Finally we renamed some columns for easier referencing later on.

### EDA
We further streamlined the dataset by dropping a few more columns after cross referencing them with each other. We plotted some columns against the 2 revenue columns and found some discrepancies, made some assumptions, corrected them and created a few columns for analysis purpose. This is a preliminary analysis. A more detailed analysis will be done on Power BI. Data is prepped and ready for modeling and Power BI analysis.

### Modeling
For this analysis, we chose an ARIMA model to conduct a time series forecast to predict rooms sold/demand over a daily interval using the series past values. It is important to note that there is a steep climb at the first part and a sharp dip in the last part of this dataset. After many rounds of experimenting with different train/test split ratio, due to the nature of this particular dataset, we have come to the conclusion that the sharp increment/decrement in the data made it relatively unfriendly to conventional data science machine learning modeling. The machine tend to overfit on the training data and the predictions are heavily skewed. With that, we declare that a deployable model is not achievable. On that note, we do offer a model without the sharp increment/decrement in the data and to make a forecast with a 95% confidence interval. 

### Business analysis
**Overview**


From the Power BI deck, we see that this analysis is conducted on a dataset over a period of approximately 14 months from July 2012 to August 2013. There are a total of 82 hotels operating in 23 cities across the nation China. The chain generated a total of 10.09m USD during this period, out of which 7.74m USD is from room revenue and the remainder from other sources like room service and spa etc. One thing we note is that the check in counts and revenue in each city stayed pretty consistent throughout this period with not much changes or sudden jumps which means the business is operationally stable.


**Metrics and rules**


For this analysis, we will be using check in counts and revenue as the metrics. We will also be employing the Pareto Principle, also known as the 80/20 rule, which asserts that, in our case, 80% of our revenue come from 20% of our product(hotel).


**Analysis by City**


Going by the Pareto Principal, we will focus in on the top 6 hotels (20% of 23 cities rounded up + 1 to cover a slightly wider range). We calculated the average total revenue per city to be around 439,000 USD and average check in count per city to be 4479. Top 6 Cities by total revenue are Shanghai, Hangzhou, Suzhou, Nanjing, Guangzhou and Jiuzhaigou and top 6 cities by check in count are basically the same top 5 but with Wuxi coming in at the 6th position. One reason why these cities are making it into the tables is because they have more hotels so generally it makes sense that more hotels means more check ins and higher revenue and that brings our attention to 5 cities: Jiuzhaigou and Wuxi for good reasons and Chengdu, Shenzhen and Chongqing for bad ones. 


With 2 hotels in the city, while Jiuzhaigou is 57% below the nation's average in check in counts, they did very well in bringing in revenue to rank at 6th. On the other hand, Wuxi, with only 1 hotel, while missing the total revenue average by around 29%, managed to boost a high check in count to rank, again, at 6th place. It is worth zooming into these 2 cities to find out more about how and what they have been doing to bring in good results like they have. 


According to the Chinese city tier system by Yicai global 2017, Shenzhen is a tier 1 city and Chengdu and Chongqing are new tier 1 (used to be tier 2 in the older system) cities. With 5 hotels and being a tier 1 city, it is worrying to see Shenzhen performing below nation's average in both metrics, specifically near 30% below for both. Similarly, being a new tier 1 city, with 6 hotels in Chengdu and 4 in Chongqing, these 2 cities disappointedly did 45 to 67% below average in both departments.


**Cities to focus**

| City          	| Reason                           	|
|----------------	|----------------------------------	|
| **Jiuzhaigou** 	| High revenue low check in counts 	|
| **Wuxi**       	| High check in counts low revenue 	|

| City          	| Reason                           	|
|---------------	|----------------------------------	|
| **Shenzhen**  	| Low in revenue & check in counts 	|
| **Chengdu**   	| Low in revenue & check in counts 	|
| **Chongqing** 	| Low in revenue & check in counts 	|

**Analysis by hotel**


Using the same 20% rule, we will focus in on the top 17 performing hotels. A simple comparison showed us that the top 17 hotels by total revenue are the same as the ones by check in counts which made perfect sense because these 2 metrics are closely correlated where most of the revenue are generated from selling rooms. All hotels except Suzhou City Rodman, Hangzhou Grand Jordan and Shanghai Grant V performed above national average on both metrics. Those 3 fared slightly below in the revenue section by between 7 to 40%. 


Having acknowledged the correlation between check in counts and revenue, there are two categories we are particularly interested in: low check in counts high revenue and high check in counts low revenue. In the first category, we have Shanghai Urban Pippen, Hanzhou Pippen, Shanghai East Rodman and Shanghai City Pippen. These 5 hotels brought in relatively high revenue with a comparatively low check in counts. 2 points of interest may be how they managed that and also how to increase the check in counts.


In the second category we have Guangzhou Urban Jordan, Shanghai Grant I and Wuxi Pippen. Likewise from the previous category, it would be interesting to find out why these 3 hotels are generating relatively low revenue when so many rooms were sold and also on the positive end, dive deeper into how they managed to sell so many rooms.


**Hotels to focus**

| Revenue below average     	|
|---------------------------	|
| **Suzhou City Rodman**    	|
| **Hangzhou Grand Jordan** 	|
| **Shanghai Grant V**      	|

| Low check in counts high revenue 	|
|----------------------------------	|
| **Shanghai Urban Pippen**        	|
| **Hangzhou Pippen**              	|
| **Shanghai East Rodman**         	|
| **Shanghai City Pippen**         	|

| High check in counts low revenue 	|
|----------------------------------	|
| **Guangzhou Urban Jordan**       	|
| **Shanghai Grant I**             	|
| **Wuxi Pippen**                  	|

**Analysis by member**


We generated a Decomposition Tree on the page Revenue decomposition to provide some analysis by members. Using this tree, we can drill down from city to individual hotel to every member who have stayed there and which city they are from. This can be useful for the Sales & marketing team for target precision marketing purpose. It is also shown that our chain is generating 75% of revenue from Platinum members and the remaining 25% from gold members,  as well as 68% from tier 1, 31% from tier 2 and less than 1% of revenue from tier 3 members. This is good evidence that our loyalty program 'The Staycation Card' is working well and we should work and further capitalize on it.


**The decline**

Addressing the sudden and sharp dip in check in counts starting from July 2013, we went ahead to conduct some research and found two possibilities.
- 2013 Southwest flood
- Xi Jinping's ascension to power

In July 2013, much of southwest China experienced heavy rainfall that led to widespread flooding. Sichuan was the hardest hit. An estimated 6 million lives were disrupted by the floods. Following that, typhoons and storms made landfall over China devastating northern, northeastern and southern provinces affecting, in total, an estimated 72 million lives. As we can see on our deck, that is EXACTLY how our hotels are distributed across the country. This natural disaster probably single handedly destroyed the sales during this period forcing most, if not all of bookings to be canceled. 


The second reason could be due to Xi Jinping's ascension to power in March 2013. Upon taking office, Xi launched a far reaching campaign against corruption where he outlined the 'eight-point guide' listing rules intended to curb corruption and waste during official party business including traveling only when necessary and reducing entourage to bare minimum. The campaign immediately went into strict enforcement causing many booked trips to be canceled or reduced in size. While we understand that our group is focused more on staycation than business traveling market, this was a major event during that time which hugely affected hotel industry. Even if it did not have a direct impact on us, there could still be some degree of indirect effect.

**Recommendations**


While it is impossible to control natural disasters, we can
1. Change marketing tactics
- Storms and typhoons are mostly seasonal and weather stations usually issue warnings before they happen. While dealing with a natural disaster, we can avoid waste in marketing by pausing marketing campaigns, replacing regular email campaigns with messages crafted for an emergency and anticipate cancellations.
2. Regulate updates
- We can maximize our social media platforms to offer updates to guests, concerned parties and community and keep guests and interested parties notified on the progress of the events.
3. Recovery
- Ensuring safety of guests and employees should be primary concern. If possible, keep track of any damages during the event to better formulate cleanup plans and focus on recovery after the event for reopening.
4. Diversification
- To prevent being affected too much by a certain group (in this case, business travelers), we can diversify and target different market groups like family or couple so we are not dependent on any specific groups.


## Data dictionary

| Glossary     	|                                                                                                                    	|
|--------------	|--------------------------------------------------------------------------------------------------------------------	|
|              	|                                                                                                                    	|
| Night        	| One paid-for night at a hotel.                                                                                     	|
| Stay         	| A members stay at a hotel.    One stay can be made up of multiple nights.                                          	|
| Booking      	| A booking made by a member for a hotel stay.  Note that a booking may be cancelled or the   member may not show.   	|
| Actual stay  	| A stay that is actually made (consumed) by a member (as   opposed to a booking).                                   	|
| Room revenue 	| Revenue from the room booking portion of a member's stay   (excluding food, drinks, room service, telephone, etc.) 	|
| ADR          	| Average Daily Rate.    Average income per occupied room.                                                           	|
| E-mailable   	| Loyalty members whose e-mail address is known to the client   (for marketing purposes).                            	|
| SMS-able     	| Loyalty members whose mobile phone number is known to the   client (for marketing purposes).                       	|
| Lead time    	| The period of time between a member's date of booking and   planned data of stay.                                  	|

## Data Cleaning

In [1]:
# Import libraries
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

In [2]:
# Read in data
xls = pd.ExcelFile(
    r'C:\Users\user\Rough_paper\project_capstone\data\Planet Jordan Hotels_Raw Data_V2.xlsx')
data = pd.read_excel(xls, 'Raw Data')

In [3]:
# Checking loaded data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103152 entries, 0 to 103151
Data columns (total 28 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   new_mbr_id           103152 non-null  int64         
 1   MBR_TYP              103152 non-null  object        
 2   em                   103152 non-null  object        
 3   sms                  103152 non-null  object        
 4   mem_city             103152 non-null  object        
 5   MEM_Tier             103152 non-null  int64         
 6   BOOKING_DT           103152 non-null  datetime64[ns]
 7   BKG_MNEMONIC_CD      103152 non-null  object        
 8   CONF_DT              103134 non-null  datetime64[ns]
 9   lead_time            101136 non-null  float64       
 10  CAL_DAY_OF_WEEK      921 non-null     object        
 11  CK_IN_DT             103152 non-null  datetime64[ns]
 12  CK_OUT_DT            103152 non-null  datetime64[ns]
 13  RM_TYP        

In [ ]:
data.head()

In [4]:
# Reformat column names to lower case for easier handling
data.columns = data.columns.str.lower()

In [5]:
# Check for null values
null_col = data.isnull().sum()
null_col[null_col > 0]

conf_dt                18
lead_time            2016
cal_day_of_week    102231
rm_typ                112
yrmonth            102231
dtype: int64

In [6]:
# Check for unusual values in categorical features
cat_feat = data.select_dtypes(include=['object'])
for column in cat_feat.columns:
    print(column)
    print(cat_feat[column].value_counts(dropna=False))
    print('\n')

mbr_typ
PLTN    77504
GOLD    25648
Name: mbr_typ, dtype: int64


em
Y    92848
N    10304
Name: em, dtype: int64


sms
Y    94192
N     8960
Name: sms, dtype: int64


mem_city
SHANGHAI     58688
GUANGZHOU    12992
NANJING       9184
HANGZHOU      4256
CHONGQING     2800
CHENGDU       2576
SUZHOU        2240
WUXI          2128
SHENZHEN      2016
NINGBO        1568
CHANGZHOU     1456
ZHOUHAI        896
FOSHAN         896
HUIZHOU        560
XIAMEN         560
NANTONG        224
KUNMING        112
Name: mem_city, dtype: int64


bkg_mnemonic_cd
CNCRO    103152
Name: bkg_mnemonic_cd, dtype: int64


cal_day_of_week
NaN         102231
SATURDAY       402
FRIDAY         376
SUNDAY         140
MONDAY           2
TUESDAY          1
Name: cal_day_of_week, dtype: int64


rm_typ
KEXN    9072
OQNN    7392
KFTN    6272
TTWN    4368
KCIG    4144
        ... 
TPLG     112
TLON     112
OSBS     112
XLOG     112
NaN      112
Name: rm_typ, Length: 81, dtype: int64


chain_cd
The Pippen    28560
The Rodman 

In [7]:
# checking for unusual values in htl_nm just to be sure
pd.set_option('display.max_rows', 500)
data['htl_nm'].value_counts()

Guangzhou Urban Jordan    6608
Suzhou Pippen             5264
Nanjing Pippen            5040
Shanghai Grant I          4816
Wuxi Pippen               4368
Shanghai Urban Jordan     3696
Suzhou City Rodman        3248
Shanghai Urban Pippen     3248
Hangzhou Pippen           3024
Suzhou Urban Rodman       2800
Hangzhou City Rodman      2800
Shanghai City Pippen      2800
Nanjing Rodman            2464
Shanghai East Rodman      2240
Nanjing Jordan            2240
Shanghai Grant V          2128
Shanghai Grant III        2016
Hangzhou Grand Jordan     2016
Jiuzhai Pippen            1904
Shanghai Grand Kerr       1792
Shenzhen Grant II         1680
Shanghai Rural Rodman     1568
Shanghai Urban Rodman     1568
Hangzhou Royal Jordan     1568
Hangzhou Urban Rodman     1456
Suzhou Royal Jordan       1456
Ningbo Jordan             1456
Shanghai West Rodman      1232
Shanghai Grant VI         1232
Chongqing Grant           1008
Guangzhou City Rodman     1008
Hangzhou Jordan Resort    1008
Changshu

**Taking reference from the hotel city column, we would assume 'Zhuzai Rodman' should be 'Zhuhai Rodman' and 'Maashan Jordan' should be 'Maanshan Jordan' so we will rename the entries**

In [8]:
data['htl_nm'] = data['htl_nm'].replace(
    {'Zhuzai Rodman': 'Zhuhai Rodman', 'Maashan Jordan': 'Maanshan Jordan'})

In [9]:
# checking for unusual values in rm_typ just to be sure
pd.set_option('display.max_rows', 500)
data['rm_typ'].value_counts()

KEXN    9072
OQNN    7392
KFTN    6272
TTWN    4368
KCIG    4144
KFTS    4032
KFTG    3472
XEXG    3472
XSTN    3360
XSTS    3024
TFTN    3024
KNGN    2688
KDXG    2688
TEXN    2688
XSTG    2128
OQNS    2128
TCIG    2016
TDBN    1904
TSUG    1680
XEXN    1680
KVUN    1568
XKLG    1568
XEXS    1456
TFTS    1344
KVUG    1232
KEXS    1232
XBUG    1120
KNGS    1120
KSUG    1008
KBUG    1008
KLOG     896
XFTS     896
XKLN     896
KEEN     896
XFTN     784
TDXG     784
XVUN     784
KSBN     784
TFTG     672
XOTN     672
XVUG     672
TOTN     672
OVUN     560
KLON     560
OEXN     560
KVUS     560
KOTN     560
XOTS     448
TEXS     448
XFTG     448
TTWS     336
KEXG     336
KOTS     336
XVUS     336
KEES     336
TDBS     224
TSBN     224
TVUN     224
XLON     224
TEES     224
XOTG     224
XKLS     224
KLOS     224
KPLG     224
OSNN     224
OFTS     112
XLOG     112
CSPG     112
TLON     112
OSBS     112
TSTG     112
OVUS     112
XDBS     112
CSHN     112
CDXG     112
OLON     112
OFTN     112

**Confirmed that there are 112 null values in rm_typ column and no unusual values in other categorical columns**

In [10]:
#Check for abnormal values in numerical features
data.describe().T

count          mean         std           min  \
new_mbr_id           103152.0  1.000005e+08  265.870931  1.000000e+08   
mem_tier             103152.0  1.293160e+00    0.471614  1.000000e+00   
lead_time            101136.0  6.097453e+00    8.640609  0.000000e+00   
yrmonth                 921.0  2.012086e+05    0.525671  2.012070e+05   
mnemonic_cd          103152.0  3.365581e+01   25.226435  1.000000e+00   
hotel_tier           103152.0  1.575461e+00    0.571688  1.000000e+00   
count_as_stay        103152.0  9.978284e-01    0.046550  0.000000e+00   
count_as_bkg         103152.0  8.393051e-01    0.367251  0.000000e+00   
nt_qty               103152.0  1.224756e+00    0.513105  1.000000e+00   
rm_rev_amt           103152.0  4.781189e+02  625.464408  0.000000e+00   
rm_rev_usd_amt       103152.0  7.525922e+01   98.494716  0.000000e+00   
rm_rev_usd_bdgt      103152.0  7.527686e+01   98.475489  0.000000e+00   
tot_sty_rev_usd_amt  103152.0  8.776810e+01  115.850592 -5.414000e+01   

                              25%           50%           75%           max  
new_mbr_id           1.000002e+08  1.000005e+08  1.000007e+08  1.000009e+08  
mem_tier             1.000000e+00  1.000000e+00  2.000000e+00  3.000000e+00  
lead_time            1.000000e+00  3.000000e+00  8.000000e+00  4.500000e+01  
yrmonth              2.012080e+05  2.012090e+05  2.012090e+05  2.012100e+05  
mnemonic_cd          1.300000e+01  3.000000e+01  5.000000e+01  9.500000e+01  
hotel_tier           1.000000e+00  2.000000e+00  2.000000e+00  3.000000e+00  
count_as_stay        1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  
count_as_bkg         1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  
nt_qty               1.000000e+00  1.000000e+00  1.000000e+00  7.000000e+00  
rm_rev_amt           1.782000e+02  3.240000e+02  6.000000e+02  9.186000e+03  
rm_rev_usd_amt       2.801000e+01  5.104000e+01  9.451000e+01  1.446960e+03  
rm_rev_usd_bdgt      2.805648e+01  5.101179e+01  9.446628e+01  1.446279e+03  
tot_sty_rev_usd_amt  2.817000e+01  5.882000e+01  1.191900e+02  1.712750e+03

**Most of yrmonth and 2016 values in lead_time are null and the rest of the data looks to be normal**

In [11]:
data.sample(10)

new_mbr_id mbr_typ em sms  mem_city  mem_tier booking_dt  \
86542    100000890    PLTN  Y   Y  SHANGHAI         1 2013-01-07   
47371    100000401    PLTN  Y   Y  SHANGHAI         1 2013-01-22   
8725     100000437    PLTN  N   Y  HANGZHOU         2 2013-02-01   
92701    100000602    PLTN  Y   Y  SHANGHAI         1 2012-07-13   
25068    100000202    GOLD  Y   Y  HANGZHOU         2 2013-02-14   
34390    100000314    PLTN  Y   Y  SHANGHAI         1 2013-04-29   
6871     100000425    PLTN  Y   Y  SHANGHAI         1 2013-06-08   
102028   100000719    PLTN  Y   Y  SHANGHAI         1 2013-01-18   
87766    100000272    PLTN  Y   Y  SHENZHEN         1 2012-11-24   
73192    100000434    PLTN  N   Y  HANGZHOU         2 2012-11-29   

       bkg_mnemonic_cd    conf_dt  lead_time  ...                 htl_nm  \
86542            CNCRO 2013-01-14        4.0  ...   Shanghai City Pippen   
47371            CNCRO 2013-02-05        0.0  ...   Shanghai City Pippen   
8725             CNCRO 2013-02-02        4.0  ...  Shanghai Urban Rodman   
92701            CNCRO 2012-07-28        0.0  ...         Nanjing Pippen   
25068            CNCRO 2013-02-22       17.0  ...          Ningbo Jordan   
34390            CNCRO 2013-05-13       42.0  ...          Suzhou Pippen   
6871             CNCRO 2013-06-21        0.0  ...   Shanghai East Rodman   
102028           CNCRO 2013-02-01        3.0  ...   Hangzhou City Rodman   
87766            CNCRO 2012-12-02        3.0  ...  Guangzhou City Rodman   
73192            CNCRO 2012-12-06        1.0  ...  Hangzhou Grand Jordan   

       hotel_city hotel_tier count_as_stay  count_as_bkg nt_qty  rm_rev_amt  \
86542    SHANGHAI          1             1             1      2      921.60   
47371    SHANGHAI          1             1             1      1        0.00   
8725     SHANGHAI          1             1             1      1      322.44   
92701     NANJING          2             1             1      1      458.00   
25068      NINGBO          2             1             1      2      575.32   
34390      SUZHOU          2             1             1      1      660.00   
6871     SHANGHAI          1             1             0      1      729.85   
102028   HANGZHOU          2             1             0      2        0.00   
87766   GUANGZHOU          1             1             1      1      592.80   
73192    HANGZHOU          2             1             1      1      192.00   

       rm_rev_usd_amt rm_rev_usd_bdgt tot_sty_rev_usd_amt  
86542          144.86      145.100200              167.74  
47371            0.00        0.000000                0.00  
8725            50.79       50.766177               59.20  
92701           71.99       72.109257               86.56  
25068           90.62       90.580563              116.82  
34390          103.96      103.912903              138.10  
6871           114.96      114.910352              157.07  
102028           0.00        0.000000                0.00  
87766           93.18       93.332680              107.16  
73192           30.24       30.229208               42.66  

[10 rows x 28 columns]

| Glossary     	|                                                                                                                    	|
|--------------	|--------------------------------------------------------------------------------------------------------------------	|
|              	|                                                                                                                    	|
| Night        	| One paid-for night at a hotel.                                                                                     	|
| Stay         	| A members stay at a hotel.    One stay can be made up of multiple nights.                                          	|
| Booking      	| A booking made by a member for a hotel stay.  Note that a booking may be cancelled or the   member may not show.   	|
| Actual stay  	| A stay that is actually made (consumed) by a member (as   opposed to a booking).                                   	|
| Room revenue 	| Revenue from the room booking portion of a member's stay   (excluding food, drinks, room service, telephone, etc.) 	|
| ADR          	| Average Daily Rate.    Average income per occupied room.                                                           	|
| E-mailable   	| Loyalty members whose e-mail address is known to the client   (for marketing purposes).                            	|
| SMS-able     	| Loyalty members whose mobile phone number is known to the   client (for marketing purposes).                       	|
| Lead time    	| The period of time between a member's date of booking and   planned data of stay.                                  	|

**According to the glossary given, lead time is supposed to be number of days between booking and check in date but we found that the data given in this column has too many errors so we will drop the column for now. If need be, we can always create the column later on and compute the values on our own using booking and check in date**

### Handling null values

**There are 5 columns with null values**
- conf_dt                18
- lead_time            2016
- cal_day_of_week    102231
- rm_typ                112
- yrmonth            102231

**We will drop cal_day_of_week and yrmonth for having more than 90% of values as null. While we're at it, we will drop bkg_mnemonic_cd for having the same value for whole column and lead_time for reason mentioned above**

In [12]:
data.drop(columns=['cal_day_of_week', 'yrmonth',
          'bkg_mnemonic_cd', 'lead_time'], inplace=True)

**There are 18 null values in conf_dt which is less than 0.02% of the data and 112 in rm_typ which is about 0.1% of the data, we will just drop the rows**

In [13]:
data.dropna(axis=0, how='any', inplace=True)

In [14]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 103022 entries, 0 to 103151
Data columns (total 24 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   new_mbr_id           103022 non-null  int64         
 1   mbr_typ              103022 non-null  object        
 2   em                   103022 non-null  object        
 3   sms                  103022 non-null  object        
 4   mem_city             103022 non-null  object        
 5   mem_tier             103022 non-null  int64         
 6   booking_dt           103022 non-null  datetime64[ns]
 7   conf_dt              103022 non-null  datetime64[ns]
 8   ck_in_dt             103022 non-null  datetime64[ns]
 9   ck_out_dt            103022 non-null  datetime64[ns]
 10  rm_typ               103022 non-null  object        
 11  chain_cd             103022 non-null  object        
 12  mnemonic_cd          103022 non-null  int64         
 13  hotel_cluster 

In [15]:
# renaming column names for clarity
column = [0, 1, 2, 14, 19]
new_names = ['mem_id', 'mem_type', 'email', 'hotel_name', 'nights_stayed']
old_names = data.columns[column]
data.rename(columns=dict(zip(old_names, new_names)), inplace=True)

In [16]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 103022 entries, 0 to 103151
Data columns (total 24 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   mem_id               103022 non-null  int64         
 1   mem_type             103022 non-null  object        
 2   email                103022 non-null  object        
 3   sms                  103022 non-null  object        
 4   mem_city             103022 non-null  object        
 5   mem_tier             103022 non-null  int64         
 6   booking_dt           103022 non-null  datetime64[ns]
 7   conf_dt              103022 non-null  datetime64[ns]
 8   ck_in_dt             103022 non-null  datetime64[ns]
 9   ck_out_dt            103022 non-null  datetime64[ns]
 10  rm_typ               103022 non-null  object        
 11  chain_cd             103022 non-null  object        
 12  mnemonic_cd          103022 non-null  int64         
 13  hotel_cluster 

In [17]:
data.to_csv('./data/data_cleaned.csv', index=False)